File: source/models/structures/sources.py
import json
from pydantic import BaseModel
from enum import Enum
from typing import List, Optional, Union


class GooglenewsFeedType(str, Enum):
    SEARCH = "search"
    TOPIC = "topic"
    TOP_TOPICS = "top_topics"
    LOCATION = "location"


class GoogleNewsConfig(BaseModel):
    feed_type: Optional[GooglenewsFeedType] = None
    lang: Optional[str] = "en"
    country: Optional[str] = "US"
    topic: Optional[Union[str, List[str]]] = None
    query: Optional[str] = None
    location: Optional[Union[str, List[str]]] = None
    since: Optional[str] = "1d"
    articles: Optional[int] = None

    def to_json(self):
        return json.dumps(self.model_dump(), default=str)


class HackerNewsFeedType(str, Enum):
    NEWEST = "newest"
    NEWCOMMENTS = "newcomments"
    FRONT_PAGE = "frontpage"
    BEST_COMMENTS = "bestcomments"
    ASK = "ask"
    SHOW = "show"
    POLLS = "polls"
    JOBS = "jobs"
    WHOISHIRING = "whoishiring"


class HackerNewsConfig(BaseModel):
    feed_type: HackerNewsFeedType
    query: Optional[str] = None
    points: Optional[int] = None
    comments: Optional[int] = None
    articles: Optional[int] = None

    def to_json(self):
        return json.dumps(self.model_dump(), default=str)


class TechCrunchNewsConfig(BaseModel):
    articles: Optional[int] = None


class YleFeedType(str, Enum):
    MAJOR_HEADLINES = "majorHeadlines"
    MOST_READ = "mostRead"
    TOPICS = "topics"


class YleLanguage(str, Enum):
    EN = "en"
    FI = "fi"


class YleNewsConfig(BaseModel):
    feed_type: Optional[YleFeedType] = None
    type: Optional[YleFeedType] = None
    articles: Optional[int] = None
    topics: Optional[List[str]] = None
    locations: Optional[List[str]] = None
    lang: Optional[YleLanguage] = YleLanguage.FI

File: lib_js/models/structures/sources.js
/**
 * Enumeration for Google News feed types.
 */
export const GooglenewsFeedType = {
    SEARCH: "search",
    TOPIC: "topic",
    TOP_TOPICS: "top_topics",
    LOCATION: "location"
};

/**
 * GoogleNewsConfig class representing the configuration for Google News.
 */
export class GoogleNewsConfig {
    constructor({
        feed_type = null,
        lang = "en",
        country = "US",
        topic = null,
        query = null,
        location = null,
        since = "1d",
        articles = null
    } = {}) {
        this.feed_type = feed_type;
        this.lang = lang;
        this.country = country;
        this.topic = topic;
        this.query = query;
        this.location = location;
        this.since = since;
        this.articles = articles;
    }
}

/**
 * Enumeration for Hacker News feed types.
 */
export const HackerNewsFeedType = {
    NEWEST: "newest",
    NEWCOMMENTS: "newcomments",
    FRONT_PAGE: "frontpage",
    BEST_COMMENTS: "bestcomments",
    ASK: "ask",
    SHOW: "show",
    POLLS: "polls",
    JOBS: "jobs",
    WHOISHIRING: "whoishiring"
};

/**
 * HackerNewsConfig class representing the configuration for Hacker News.
 */
export class HackerNewsConfig {
    constructor({
        feed_type,
        query = null,
        points = null,
        comments = null,
        articles = null
    } = {}) {
        this.feed_type = feed_type;
        this.query = query;
        this.points = points;
        this.comments = comments;
        this.articles = articles;
    }
}

/**
 * TechCrunchNewsConfig class representing the configuration for TechCrunch News.
 */
export class TechCrunchNewsConfig {
    constructor({ articles = null } = {}) {
        this.articles = articles;
    }
}

/**
 * Enumeration for Yle feed types.
 */
export const YleFeedType = {
    MAJOR_HEADLINES: "majorHeadlines",
    MOST_READ: "mostRead",
    TOPICS: "topics"
};

/**
 * Enumeration for Yle languages.
 */
export const YleLanguage = {
    EN: "en",
    FI: "fi"
};

/**
 * YleNewsConfig class representing the configuration for Yle News.
 */
export class YleNewsConfig {
    constructor({
        feed_type = null,
        type = null,
        articles = null,
        topics = null,
        locations = null,
        lang = YleLanguage.FI
    } = {}) {
        this.feed_type = feed_type;
        this.type = type;
        this.articles = articles;
        this.topics = topics;
        this.locations = locations;
        this.lang = lang;
    }
}
